cpp/src/arrow/python/numpy_to_arrow.cc

```diff
@@ -477,7 +477,7 @@ inline Status NumPyConverter::ConvertData<Date32Type>(std::shared_ptr<Buffer>* d
   } else {
     RETURN_NOT_OK(NumPyDtypeToArrow(reinterpret_cast<PyObject*>(dtype_), &input_type));
     if (!input_type->Equals(*type_)) {
-      RETURN_NOT_OK(CastBuffer(input_type, *data, length_, nullptr, 0, type_,
+      RETURN_NOT_OK(CastBuffer(input_type, *data, length_, null_bitmap_, null_count_, type_,
```

Comment:
@wesm I can't write failing test cases for the date types, even with explicitly passing a mask.

```
pa.Array.from_pandas(
    np.array([17259, -1, 17260, 17261], dtype='int32'), 
    mask=np.array([False, True, False, False]), 
    type=pa.int32(), 
    safe=True
)
```

Reply:
I don't think it's possible to reach this code path right now. the closest I got was

```
arr = pa.array(np.array([0, 0, 1, 0], dtype='int64'),
               mask=[False, False, True, False],
               type='date32')
## -- End pasted text --
---------------------------------------------------------------------------
ArrowNotImplementedError                  Traceback (most recent call last)
<ipython-input-5-3dc2053fe073> in <module>
      1 arr = pa.array(np.array([0, 0, 1, 0], dtype='int64'),
      2                mask=[False, False, True, False],
----> 3                type='date32')

~/code/arrow/python/pyarrow/array.pxi in pyarrow.lib.array()
    169             values, type = pdcompat.get_datetimetz_type(values, obj.dtype,
    170                                                         type)
--> 171             return _ndarray_to_array(values, mask, type, from_pandas, safe,
    172                                      pool)
    173     else:

~/code/arrow/python/pyarrow/array.pxi in pyarrow.lib._ndarray_to_array()
     78 
     79     with nogil:
---> 80         check_status(NdarrayToArrow(pool, values, mask, from_pandas,
     81                                     c_type, cast_options, &chunked_out))
     82 

~/code/arrow/python/pyarrow/error.pxi in pyarrow.lib.check_status()
     87             raise ArrowKeyError(message)
     88         elif status.IsNotImplemented():
---> 89             raise ArrowNotImplementedError(message)
     90         elif status.IsTypeError():
     91             raise ArrowTypeError(message)

ArrowNotImplementedError: ../src/arrow/python/numpy_to_arrow.cc:803 : ../src/arrow/python/numpy_to_arrow.cc:285 : ../src/arrow/python/numpy_to_arrow.cc:481 : ../src/arrow/python/numpy_to_arrow.cc:349 : ../src/arrow/compute/kernels/cast.cc:1436 : ../src/arrow/compute/kernels/cast.cc:1423 : No cast implemented from int64 to date32[day]
```

Note the annoying ARROW-4252 bug

Reply:
Ok, fixing the pr

Reply:
Added a TODO note